Webpack Upgrade to 4.5.0

armaanahluwalia opened the pull request:
This commit involves upgrading webpack to the latest version
at the time of authoring. It also involves upgrading webpack
loaders and types and adding webpack-cli as a dependency for
running webpack from the shell.

timabbott commented:
@armaanahluwalia it looks like this is failing in CI.  See if you can reproduce with `tools/build-release-tarball` -- that's the job that's failing.

cPhost commented:
Despite the CI Failure this is still not going to work. The dev server (production server too) is broken, the reason being the `expose-loader` not working (FWIW it does work with v3.x also).

@armaanahluwalia if you are familiar with webpack, you can try to make it work with `webpack.providePlugin` or something like that. I doubt we will find another package as a replacement of `expose-loader` reason being webpack modular way of bundling.

armaanahluwalia commented:
@timabbott Sorry about that I updated the PR with the change. Problem was that webpack-cli wasn't available as an executable since it was a devDependency so i moved it as a dependency instead.

@cPhost Thanks for the input here. I'm not exactly sure I follow. I've simply updated the expose-loader version to be compatible with the the latest version of webpack. Seems to be working for me locally. Let me know if I'm misunderstanding what you mean.

timabbott commented:
@armaanahluwalia can you explain what the story is with the changes to the webpack configuration you made?  E.g.
* You removed the source-map-loader related configuration.  Will that break source maps in production?
* You changed the `devtool` default
etc.

The ones of those that don't have to change as part of upgrading webpack should be their own commits with their own commit messages explaining the change that's a separate commit from the version upgrade transition (e.g. the `devtool` change seems like it belongs in this category).

If any change has to be part of the webpack upgrade commit (i.e. that change doesn't work with older webpack and is required with newer webpack), it can go in the main commit; but the commit message should explain why.

cPhost commented:
@armaanahluwalia it does work for me locally nor it did work for me when i tried to upgrade it. Here is the screenshot of all the errors, which i faced before and for this PR, where are related to expose-loader not working:
![screen shot 2018-04-17 at 8 06 35 pm](https://user-images.githubusercontent.com/23620441/38905035-f732a98e-427b-11e8-8af6-3443b5a19fbc.png)

armaanahluwalia commented:
Hi @timabbott 

As per your comments, I've split up the code changes into separate commits. 

Interesting point about the 'source-map-loader'. As per the comment that code relates to adding source-maps for TypeScript files. What I noticed is that the TypeScript compilation was actually broken in the earlier webpack configuration and since there were no TypeScript files in the codebase it wasn't affecting anything. Anyway, I went ahead and fixed the configuration and tested that TypeScript files can be compiled by webpack. Also, regarding the code for source-map-loader as per [this link](https://github.com/webpack/webpack.js.org/issues/1257) that code should no longer be required. In any case I would like to verify that it isn't indeed breaking anything for the production build. Is there a good way to test this while developing? I checked that third-party libraries and TypeScript files have source maps locally.

@priyankp10 Thanks for the input. It seems to be working fine for me. Let me know if you see anything else.

timabbott commented:
@priyankp10 how did you do your testing?  This seems to work for me as well.

@aero31aero can you try test-deploying this on nightly to see how it works in production?

priyankp10 commented:
I did it on the local development server, just did provision to install the new version, may something wrong just on my side.

Also, this PR needs a provision version bump in `version.py`

aero31aero commented:
Deployed this to https://nightly.zulipdev.org/

priyankp10 commented on tools/webpack:
```diff
@@ -21,15 +21,15 @@ STATIC_PATH = 'static/'
     """watches and rebuilds on changes, serving files from memory via webpack-dev-server"""
     webpack_args = ['node', 'node_modules/.bin/webpack-dev-server']
-    webpack_args += ['--config', 'tools/webpack.config.ts', '--watch-poll', '--port', port, "--host", host]
+    webpack_args += ['--config', 'tools/webpack.config.ts', '--progress', '--port', port, "--host", host]
```
It great that the `watch-poll` flag was removed since its handled on the webpack config side but do we really want the progress flag? All it does right now is just `10% building modules 5/5 modules 0 active` and `78 modules`; just adds noise when running dev server.

timabbott replied:
Yeah, agreed, it felt noisy to me in the `run-dev.py` output too.

armaanahluwalia replied:
Sure I'll remove the flag.

timabbott commented:
@aero31aero great!  I think we'll want to test the traceback source map piece; you might be able to do that using `blueslip.report_error("foo")` on the JS console.  (You might need to enable `BROWSER_ERROR_REPORTING` in settings first)

timabbott commented:
I tested this locally with some care, removed the addition of `--progress` and merged this, thanks @armaanahluwalia!  

@priyankp10 I think you have some PRs that we should now rebase?

timabbott commented:
@priyankp10 I should note that I also wasn't able to reproduce the issues with expose-loader, which we use for critical libraries like jQuery; can you try re-verifying?

priyankp10 commented:
I did some further debugging, the reason it was not working was related to host checking of webpack, something related specifically to cloud9 (I think I tested that on droplet but can't recall correctly what I did to confirm this but...).

I found this out when I looked at, `common.js` its content was `Invalid Host Header` which was suppose to have jQuery, blueslip and other stuff.  Hence all the errors in screenshot above. Other files were working properly since we load then using the `django-pipline`

If it worked for other that means it not `expose-loader`'s fault it works, so this upgrade is fine.
